worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,136 @@
+
+  def createNewConnectionProvider(): Option[HikariCpPoolingConnectionProvider] = {
+    if(dataSource != null) {
+      Some(new HikariCpPoolingConnectionProvider(
+        props.getProperty("org.quartz.dataSource." + dataSource + ".driver"),
+        jdbcURL,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".user"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".password"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".maxConnections").toInt,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".validationQuery")
+      ))
+    } else {
+      logger.info("No job store found in config to get connections")
+      None
+    }
+  }
+
+  /**
+   * HikariCP connection pools don't automatically close when IP addresses for a hostname change. This function returns True, iff at
+   * least one of the following conditions is met:
+   *   - IP addresses have changed for the CNAME record used for DNS lookup
+   *   - causeFailoverEveryConnection is set to "true", which is used for testing failover functionality
+   *
+   * @param pool
+   *   the connection pool currently being used
+   * @param dnsIP
+   *   the IP returned when performing a DNS lookup
+   * @return
+   */
+  private def hasIpAddressChanged(pool: Pool, dnsIP: String): Boolean = {
+    causeFailoverEveryConnection || pool.ip != dnsIP
+  }
+
+  @tailrec
+  private def retryGettingIp(n: Int)(fn: => String): String = {
+    try {
+      return fn
+    } catch {
+      // Failed to resolve it from JVM
+      case e: UnknownHostException if n > 1 =>
+    }
+    // Wait 10 milliseconds between retries
+    Thread.sleep(10)
+    retryGettingIp(n - 1)(fn)
+  }
+
+  def _getIp: String = {
+    retryGettingIp(getIpNumRetries) {
+      // Get the ip address of the hostname. The result is cached in the JVM
+      val ip = java.net.InetAddress.getByName(dataSourceHostname).getHostAddress
+      cachedIpWithExpiration.put(dataSourceHostname, CachedIpWithExpiration(ip, System.currentTimeMillis() + cachedIpTTL))
+      ip
+    }
+  }
+
+  def getIP: String = {
+    val optionCachedValue = cachedIpWithExpiration.get(dataSourceHostname)
+    if (optionCachedValue.nonEmpty) {
+      val cachedValue = optionCachedValue.get
+      if (System.currentTimeMillis() > cachedValue.expiration) {
+        _getIp
+      } else {
+        cachedValue.ip
+      }
+    } else {
+      _getIp
+    }
+  }
+
+  def getConnection = {
+    if (detectIpAddressFailover && dataSourceHostname != null) {
+      // If the IP has changed, then we know a failover has occurred, and we need to create a new hikari config
+      val newIP: String = getIP
+      if (hasIpAddressChanged(pool, newIP)) {
+        // A failover has occurred, so we evict connections softly. New connectons look up the new IP address
+        pool.connectionProvider.get.getDataSource().getHikariPoolMXBean().softEvictConnections()
```

Comment:
Instead of calling `.get` on `connectionProvider` can we map over it? It doesn't seem great to be calling `.get` on an option

Reply:
I think that's worth doing here